### ISSUE TITLE
when declaration causes pipeline to hang if input channel empty

### ISSUE THREAD

**Issue by huguesfontenelle** [nextflow-io/nextflow#1768]
### Expected behavior and actual behavior

When using the `when` declaration and the condition evaluate to false, the process is skipped.
I _expect_ the process to be skipped, and the pipeline to continue, whether or not the input channels are empty.
The _actual behaviour_ is that the pipeline freezes if one input channel is empty for a process which is otherwise skipped.

### Steps to reproduce the problem

```
#!/usr/bin/env nextflow

cheers = Channel.from 'olleH', 'aloH', 'oaiC', 'ruojnoB'
params.display = true

process reverse {
  input:
    val x from cheers
  output:
    file("${x}.txt") into reverse_output
  script:
    """
    echo ${x} | rev > ${x}.txt
    """
}

process display {
  echo true
  input:
    file(greet) from reverse_output
    val(decorate) from Channel.from(params.embelish).first()
  when:
    params.display
  script:
    """
    echo ${decorate} \$(cat ${greet}) ${decorate}
    """
}
```

### Program output 

When I specify all params (channels not empty) it works as expected, for both `--display` true or false, as long as I specify ``--embelish` as well:
```
./nextflow run main.nf --display true --embelish "===" 
N E X T F L O W  ~  version 20.07.1
Launching `main.nf` [focused_jang] - revision: d3cbfe8070
executor >  local (8)
[2d/46b407] process > reverse (2) [100%] 4 of 4 ✔
[24/aa77e0] process > display (1) [100%] 4 of 4 ✔
=== Hola ===
=== Bonjour ===
=== Ciao ===
=== Hello ===
```

If I don't want the second process to run, `--display false` but it freezes:
```
./nextflow run main.nf --display false
N E X T F L O W  ~  version 20.07.1
Launching `main.nf` [distraught_kowalevski] - revision: d3cbfe8070
[-        ] process > reverse [  0%] 0 of 2
executor >  local (4)
[db/2a3be3] process > reverse (2) [100%] 4 of 4 ✔
[-        ] process > display     -
WARN: Access to undefined parameter `embelish` -- Initialise it to a default value eg. `params.embelish = some_value`
```
I didn't specify the `embelish` parameter.

### Environment 

* Nextflow version: 20.07.1
* Java version: openjdk version "11.0.2" 2019-01-15
* Operating system: macOS
* Bash version: zsh 5.3 (x86_64-apple-darwin18.0)

### Additional context

I'm having a good amount of `when` declarations over my processes and writing dummy channels would complicate the code a lot.

**Comment by pditommaso** [comment]
The warning warned you: 

```
WARN: Access to undefined parameter `embelish` -- Initialise it to a default value eg. `params.embelish = some_value`
```

Properly declaring the params it works as expected eg 

```
params.embelish = false
```

**Comment by huguesfontenelle** [comment]
It seems that in my attempt to dumb it down to a simple script, I oversimplified it (or didn't get the essence of it). Thanks for reading though.

In real life `val(decorate) from Channel.from(params.embelish).first()` isn't a channel created from a simple parameter, but a channel either containing a file, or empty when `params.display = false`.
